hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,214 @@
+                        (totalInvocations % FULL_LOG_FREQUENCY == 0) ? log.getStackTrace() : log.getShortStackTrace()));
+            }
+        }
+    }
+
+    private SlowOperationLog getOrCreateSlowOperationLog(OperationRunner operationRunner, final Object operation) {
+        String prefix = "";
+        for (StackTraceElement stackTraceElement : operationRunner.currentThread().getStackTrace()) {
+            stackTraceStringBuilder.append(prefix).append(stackTraceElement.toString());
+            prefix = "\n";
+        }
+
+        final String stackTraceString = stackTraceStringBuilder.toString();
+        stackTraceStringBuilder.setLength(0);
+
+        return ConcurrencyUtil.getOrPutIfAbsent(slowOperationLogs, stackTraceString.hashCode(),
+                new ConstructorFunction<Integer, SlowOperationLog>() {
+                    @Override
+                    public SlowOperationLog createNew(Integer key) {
+                        return new SlowOperationLog(stackTraceString, operation);
+                    }
+                });
+    }
+
+    private boolean purge(long nowNanos, long lastLogPurge) {
+        if (nowNanos - lastLogPurge > logPurgeIntervalNanos) {
+            for (SlowOperationLog log : slowOperationLogs.values()) {
+                log.purgeInvocations(nowNanos, logRetentionNanos);
+                if (log.isEmpty()) {
+                    slowOperationLogs.remove(log.getStackTrace().hashCode());
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private static class CurrentOperationData {
+        int operationHashCode;
+        long startNanos;
+        SlowOperationLog.Invocation invocation;
+    }
+
+    private class Scanner extends Thread {
+
+        public Scanner(HazelcastThreadGroup hazelcastThreadGroup) {
+            super(hazelcastThreadGroup.getThreadNamePrefix("SlowOperationDetectorThread"));
+        }
+
+        public void run() {
+            while (running) {
+                long lastLogPurge = System.nanoTime();
+
+                while (running) {
+                    long nowNanos = System.nanoTime();
+                    long nowMillis = System.currentTimeMillis();
+
+                    scan(nowNanos, nowMillis, genericOperationRunners, genericCurrentOperationData);
+                    scan(nowNanos, nowMillis, partitionOperationRunners, partitionCurrentOperationData);
+
+                    if (purge(nowNanos, lastLogPurge)) {
+                        lastLogPurge = nowNanos;
+                    }
+
+                    try {
+                        TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
+                    } catch (Exception ignored) {
+                        EmptyStatement.ignore(ignored);
+                    }
+                }
+            }
+        }
+    }
```

Comment:
To simplify testing, I would follow the following approach:

expose a method purge. This method is the one you are going to call from your tests.

expose a method start. This method starts a thread that will call the purge method. For testing, you don't call start; so there is no thread doing stuff you don't want it to do.

This way testing is a lot easier because you don't need to deal with that nasty thread.

```
 public void purge(){
    long nowNanos = System.nanoTime();
    long nowMillis = System.currentTimeMillis();

    scan(nowNanos, nowMillis, genericOperationRunners, genericCurrentOperationData);
    scan(nowNanos, nowMillis, partitionOperationRunners, partitionCurrentOperationData);

    if (purge(nowNanos, lastLogPurge)) {
        lastLogPurge = nowNanos;
    }

 }

 public void start(){
    new ScanThread(...).start();
 }


class ScanThread extends Thread{
...
 public void run() {
                lastLogPurge = System.nanoTime();

                while (running) {
                    
                    try {
                        TimeUnit.NANOSECONDS.sleep(ONE_SECOND_IN_NANOS - (System.nanoTime() - nowNanos));
                    } catch (Exception ignored) {
                        EmptyStatement.ignore(ignored);
                    }
                }
        }
}
```